snuba/utils/streams/streaming.py

```diff
@@ -86,6 +107,287 @@ def join(self, timeout: Optional[float] = None) -> None:
+
+    def __len__(self) -> int:
+        return len(self.__batch)
+
+    def append(self, message: Message[TPayload]) -> None:
+        self.__batch.append(message)
+
+    def ready(self) -> bool:
+        if len(self.__batch) >= self.__max_batch_size:
+            return True
+        elif time.time() >= self.__deadline:
+            return True
+        else:
+            return False
+
+    def build(self) -> MessageBatch[TPayload]:
+        return self.__batch
+
+
+def parallel_transform_worker_initializer() -> None:
+    # Worker process should ignore ``SIGINT`` so that processing is not
+    # interrupted by ``KeyboardInterrupt`` during graceful shutdown.
+    signal.signal(signal.SIGINT, signal.SIG_IGN)
+
+
+def parallel_transform_worker_apply(
+    function: Callable[[Message[TPayload]], TTransformed],
+    input_batch: MessageBatch[TPayload],
+    output_block: SharedMemory,
+    start_index: int = 0,
+) -> Tuple[int, MessageBatch[TTransformed]]:
+    output_batch: MessageBatch[TTransformed] = MessageBatch(output_block)
+
+    i = start_index
+    while i < len(input_batch):
+        message = input_batch[i]
+        try:
+            output_batch.append(
+                Message(
+                    message.partition,
+                    message.offset,
+                    function(message),
+                    message.timestamp,
+                )
+            )
+        except ValueTooLarge:
+            # If the output batch cannot accept the transformed message when
+            # the batch is empty, we'll never be able to write it and should
+            # error instead of retrying. Otherwise, we need to return the
+            # values we've already accumulated and continue processing later.
+            if len(output_batch) == 0:
+                raise
+            else:
+                break
+        else:
+            i += 1
+
+    return (i, output_batch)
+
+
+class ParallelTransformStep(ProcessingStep[TPayload]):
+    def __init__(
+        self,
+        function: Callable[[Message[TPayload]], TTransformed],
+        next_step: ProcessingStep[TTransformed],
+        processes: int,
+        max_batch_size: int,
+        max_batch_time: float,
+        input_block_size: int,
+        output_block_size: int,
+    ) -> None:
+        self.__transform_function = function
+        self.__next_step = next_step
+        self.__max_batch_size = max_batch_size
+        self.__max_batch_time = max_batch_time
+
+        self.__shared_memory_manager = SharedMemoryManager()
+        self.__shared_memory_manager.start()
+
+        self.__pool = Pool(processes, initializer=parallel_transform_worker_initializer)
+
+        self.__input_blocks = [
+            self.__shared_memory_manager.SharedMemory(input_block_size)
+            for _ in range(processes)
+        ]
+
+        self.__output_blocks = [
+            self.__shared_memory_manager.SharedMemory(output_block_size)
+            for _ in range(processes)
+        ]
```

Comment:
This means there is a constant "heap" size of `processes * (input block size + output block size)`.